library/ecp_curves.c

```diff
@@ -5162,60 +5164,100 @@ static int ecp_mod_p384(mbedtls_mpi *N)
           MBEDTLS_ECP_DP_SECP384R1_ENABLED */
 
 #if defined(MBEDTLS_ECP_DP_SECP521R1_ENABLED)
-/*
- * Here we have an actual Mersenne prime, so things are more straightforward.
- * However, chunks are aligned on a 'weird' boundary (521 bits).
- */
-
 /* Size of p521 in terms of mbedtls_mpi_uint */
 #define P521_WIDTH      (521 / 8 / sizeof(mbedtls_mpi_uint) + 1)
 
 /* Bits to keep in the most significant mbedtls_mpi_uint */
 #define P521_MASK       0x01FF
 
 /*
- * Fast quasi-reduction modulo p521 (FIPS 186-3 D.2.5)
- * Write N as A1 + 2^521 A0, return A0 + A1
+ * Fast quasi-reduction modulo p521 = 2^521 - 1 (FIPS 186-3 D.2.5)
  */
 static int ecp_mod_p521(mbedtls_mpi *N)
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
-    size_t i;
-    mbedtls_mpi M;
-    mbedtls_mpi_uint Mp[P521_WIDTH + 1];
-    /* Worst case for the size of M is when mbedtls_mpi_uint is 16 bits:
-     * we need to hold bits 513 to 1056, which is 34 limbs, that is
-     * P521_WIDTH + 1. Otherwise P521_WIDTH is enough. */
+    size_t expected_width = 2 * ((521 + biL - 1) / biL);
+    MBEDTLS_MPI_CHK(mbedtls_mpi_grow(N, expected_width));
+    ret = mbedtls_ecp_mod_p521_raw(N->p, expected_width);
+cleanup:
+    return ret;
+}
 
-    if (N->n < P521_WIDTH) {
-        return 0;
-    }
+MBEDTLS_STATIC_TESTABLE
+int mbedtls_ecp_mod_p521_raw(mbedtls_mpi_uint *X, size_t X_limbs)
+{
+    mbedtls_mpi_uint carry = 0;
 
-    /* M = A1 */
-    M.s = 1;
-    M.n = N->n - (P521_WIDTH - 1);
-    if (M.n > P521_WIDTH + 1) {
-        M.n = P521_WIDTH + 1;
+    if (X_limbs > 2 * P521_WIDTH - 1) {
+        X_limbs = 2 * P521_WIDTH - 1;
+    }
+    if (X_limbs < P521_WIDTH) {
+        return 0;
```

Comment:
I was thinking that we could go more strict here:

```suggestion
    if (X_limbs != 2 * P521_WIDTH || X[P521_WIDTH - 1] !=0) {
       return MBEDTLS_ERR_ECP_BAD_INPUT_DATA;
    }
```

Requiring the input be double the modulus length in limbs is not really a constraint as the generic fixed width multiplication will produce such a result anyway. (And that is the only case where this function will be used.)

This would also allow to get rid of the if condition on line 5199 and  to do the addition in the end in one go (we would have the needed extra limb in the upper half).